File: BossFight Rest/BossFight/models/PlayerCLass.cs
using System;
using System.Collections.Generic;
using System.Linq;
using BossFight.BossFightEnums;
using BossFight.CustemExceptions;

namespace BossFight.Models
{
    public class PlayerClassRequirement
    {
        public int ClassId {get; set;}
        public int LevelRequirement {get; set;}
        public string ClassName {get; set;}

        public PlayerClassRequirement(int pClassId, int pLevelRequirement, string pClassName)
        {
            ClassId = pClassId;
            LevelRequirement = pLevelRequirement;
            ClassName = pClassName;
        }

        public override string ToString()
        {
            return $"{ClassName}:{ClassId} - lvl {LevelRequirement}";
        }
    }

    public abstract class PlayerClass
    {
        public static int DefaultHpScale = 2;
        public static int DefaultManaScale = 2;
        public static int PlayerClassId = -1;
        public static int BaseHealth = 10;
        public static int BaseMana = 10;

        public string Name {get; set;}
        public int Xp {get; set;}
        public int Level {get; set;}
        Player Player {get; set;}
        public List<WeaponType> ProficientWeaponTypesList {get; set;}
        public List<Ability> Abilities {get; set;}
        public double HpScale {get; set;} = DefaultHpScale;
        public double ManaScale {get; set;} = DefaultManaScale;
        public int MaxHp {get; set;}
        public int MaxMana {get; set;}

        // also properties ?
        public int PurchasePrice {get; set;} = 0;
        public double PlayerClassCritChance {get; set;} = 0;
        public int HpRegenRate {get; set;} = 1;
        public int ManaRegenRate {get; set;} = 1;
        public int AttackPowerBonus {get; set;} = 0;
        public int SpellPowerBonus {get; set;} = 0;
        
        protected int _classId = PlayerClassId;
        public int ClassId 
        {
            get { return _classId; } 
            set { _classId = value; }
        }

        public PlayerClass(string pName, int pXp, int pLevel, PlayerClass pPlayer, List<WeaponType> pProficientWeaponTypesList, double? pHpScale = null, double? pManaScale = null)
        {
            Name = pName;
            Xp = pXp;
            Level = pLevel;
            Player = pPlayer;
            ProficientWeaponTypesList = pProficientWeaponTypesList;
            Abilities = PrepareAvailableAbilities();
            MaxHp = 0;
            MaxMana = 0;
            if (pHpScale != null)
                HpScale = pHpScale.Value;
            if (pManaScale != null)
                ManaScale = pManaScale.Value;
            Recalculate();
        }

        public abstract Dictionary<String, Ability> PrepareAvailableAbilities();

        // fromDict
        public abstract PlayerClass FromDB(object playerClassDict, Player player);

        public static List<PlayerClassRequirement> GetClassUnlockRequirements()
        {
            return new List<PlayerClassRequirement>();
        }

        public virtual Dictionary<object, object> toDict()
        {
            var playerClassDict = new Dictionary<object, object>
            {
            };
            playerClassDict["name"] = Name;
            playerClassDict["xp"] = Xp;
            playerClassDict["level"] = Level;
            return playerClassDict;
        }

        public void Recalculate()
        {
            MaxHp = (int)Math.Floor(HpScale * Level) + BaseHealth;
            MaxMana = (int)Math.Floor(ManaScale * Level) + BaseMana;
            Abilities = PrepareAvailableAbilities();
        }

        public override string ToString()
        {
            return Name;
        }

        public virtual string ShopString(int pLengthOfLongestPlayerClassName, int pLengthOfLongestPlayerClassCost)
        {
            var purchasePriceString = String.Format("{0:n0}", PurchasePrice);
            purchasePriceString = purchasePriceString.Replace(",", ".");  //TODO is this needed?
            var playerClassRequirements = GetClassUnlockRequirements();
            var playerClassRequirementsString = String.Join(", ", (from req in playerClassRequirements select $"Level {req.LevelRequirement} {req.ClassName}"));
            return $"{Name.PadLeft(pLengthOfLongestPlayerClassName, '.')} {purchasePriceString.PadLeft(pLengthOfLongestPlayerClassCost)} gold {playerClassRequirementsString}";
        }

        public virtual string InfoString()
        {
            var classNameStr = $"Class: {Name}";
            var critChanceStr = $"Base critical chance: {PlayerClassCritChance}%";
            var baseHpManaStr = $"Start hp: {BaseHealth}; start mana: {BaseMana}";
            var scalesStr = $"{HpScale} hp per level; {ManaScale} mana per level";
            var regenStr = $"{HpRegenRate} hp per regen tick; {ManaRegenRate} mana per regen tick";
            var proficientWeaponsStr = $"Is proficient with: {String.Join(", ", ProficientWeaponTypesList.Select(p => p.ToString()))}";
            var prevLvl = Level;
            Level = 99;
            var allSpells = PrepareAvailableAbilities();
            Level = prevLvl;
            var abilitiesStr = $"Spell list:\n" + String.Join("\n", (from s in allSpells.Values select s.ToString()));
            return String.Join("\n", new List<string> {
                    classNameStr,
                    critChanceStr,
                    baseHpManaStr,
                    scalesStr,
                    regenStr,
                    abilitiesStr,
                    proficientWeaponsStr
                });
        }

        public int GetHealthRegenRate()
        {
            return HpRegenRate;
        }

        public int GetManaRegenRate()
        {
            return ManaRegenRate;
        }

        public object GetAttackPowerBonus()
        {
            return AttackPowerBonus;
        }

        public object GetSpellPowerBonus()
        {
            return SpellPowerBonus;
        }

        public void LevelUp()
        {
            Level += 1;
            Xp = 0;
            Recalculate();
            Player.RestoreAllHealthAndMana();
        }

        public virtual bool HaveMetRequirements(IEnumerable<PlayerClass> pPlayerClassList)
        {
            var requirementsMet = true;
            var classRequirements = GetClassUnlockRequirements();
            foreach (var classReq in classRequirements)
            {
                try
                {
                    pPlayerClassList.First(x => x.ClassId == classReq.ClassId);
                }
                catch (InvalidOperationException)
                {
                    throw new MyException($"You need to unlock '{classReq.ClassName}' first");
                }
                foreach (var pc in pPlayerClassList)
                {
                    if (pc.ClassId == classReq.ClassId)
                        if (pc.Level < classReq.LevelRequirement)
                            throw new MyException($"Did not meet level requirements for {classReq.ClassName}");
                }
            }
            return requirementsMet;
        }
    }

    public class Executioner : PlayerClass
    {
        public object playerClassId = 2;

        public Executioner(object player, object xp = 0, object level = 1)
            : base(xp, level, player, new List<object> {
                    WeaponType.SWORD,
                    WeaponType.MACE,
                    WeaponType.DAGGER,
                    WeaponType.POLEARM
            }, hpScale: 2.5, manaScale: 1.5)
        {
            PlayerClassCritChance = 2;
        }

        [classmethod]
        public static object FromDB(object cls, object playerClassDict = dict, object player)
        {
            var xp = playerClassDict["xp"];
            var level = Convert.ToInt32(playerClassDict["level"]);
            var playerClass = cls(player, xp, level);
            playerClass.recalculate();
            return playerClass;
        }

        public virtual object prepareAvailableAbilities()
        {
            var unlockedAbilities = new Dictionary<object, object>
            {
            };
            var sackOnHead = abilities.SackOnHead();
            unlockedAbilities[sackOnHead.magicWord] = sackOnHead;
            if (Level >= 5)
            {
                var execute = abilities.Execute();
                unlockedAbilities[execute.magicWord] = execute;
            }
            return unlockedAbilities;
        }
    }

    public class Ranger
        : PlayerClass
    {

        public object playerClassId = 3;

        public Ranger(object player, object xp = 0, object level = 1)
            : base(xp, level, player, new List<object> {
                    WeaponType.DAGGER,
                    WeaponType.THROWN,
                    WeaponType.BOW,
                    WeaponType.IMPROVISED
            })
        {
            PlayerClassCritChance = 4;
        }

        [classmethod]
        public static object FromDB(object cls, object playerClassDict = dict, object player)
        {
            var xp = playerClassDict["xp"];
            var level = Convert.ToInt32(playerClassDict["level"]);
            var playerClass = cls(player, xp, level);
            playerClass.recalculate();
            return playerClass;
        }

        public virtual object prepareAvailableAbilities()
        {
            var unlockedAbilities = new Dictionary<object, object>
            {
            };
            var doubleStrike = abilities.DoubleStrike();
            unlockedAbilities[doubleStrike.magicWord] = doubleStrike;
            if (Level >= 5)
            {
                var poisonedBait = abilities.PoisonedBait();
                unlockedAbilities[poisonedBait.magicWord] = poisonedBait;
            }
            return unlockedAbilities;
        }
    }

    public class Hexer
        : PlayerClass
    {

        public object playerClassId = 4;

        public Hexer(object player, object xp = 0, object level = 1)
            : base(xp, level, player, new List<object> {
                    WeaponType.DAGGER,
                    WeaponType.STAFF
            }, hpScale: 3.5, manaScale: 4)
        {
            PlayerClassCritChance = 3;
            PurchasePrice = 1000;
            BaseHealth = 12;
            BaseMana = 13;
            ManaRegenRate = 2;
            AttackPowerBonus = 1;
        }

        [classmethod]
        public static object FromDB(object cls, object playerClassDict = dict, object player)
        {
            var xp = playerClassDict["xp"];
            var level = Convert.ToInt32(playerClassDict["level"]);
            var playerClass = cls(player, xp, level);
            playerClass.recalculate();
            return playerClass;
        }

        public virtual object prepareAvailableAbilities()
        {
            var unlockedAbilities = new Dictionary<object, object>
            {
            };
            var hex = abilities.Hex();
            unlockedAbilities[hex.magicWord] = hex;
            if (Level >= 5)
            {
                var fractureSkin = abilities.FractureSkin();
                unlockedAbilities[fractureSkin.magicWord] = fractureSkin;
            }
            return unlockedAbilities;
        }
    }

    public class Mage
        : PlayerClass
    {

        public object playerClassId = 5;

        public Mage(object player, object xp = 0, object level = 1)
            : base(xp, level, player, new List<object> {
                    WeaponType.DAGGER,
                    WeaponType.STAFF
            }, hpScale: 3, manaScale: 4.5)
        {
            PlayerClassCritChance = 4;
            PurchasePrice = 1000;
            BaseHealth = 11;
            BaseMana = 14;
            ManaRegenRate = 2;
            SpellPowerBonus = 1;
        }

        [classmethod]
        public static object FromDB(object cls, object playerClassDict = dict, object player)
        {
            var xp = playerClassDict["xp"];
            var level = Convert.ToInt32(playerClassDict["level"]);
            var playerClass = cls(player, xp, level);
            playerClass.recalculate();
            return playerClass;
        }

        public virtual object prepareAvailableAbilities()
        {
            var unlockedAbilities = new Dictionary<object, object>
            {
            };
            var ignite = abilities.Ignite();
            unlockedAbilities[ignite.magicWord] = ignite;
            if (Level >= 5)
            {
                var enchantWeapon = abilities.EnchantWeapon();
                unlockedAbilities[enchantWeapon.magicWord] = enchantWeapon;
            }
            return unlockedAbilities;
        }
    }

    public class Barbarian
        : PlayerClass
    {

        public object playerClassId = 6;

        public Barbarian(object player, object xp = 0, object level = 1)
            : base(xp, level, player, new List<object> {
                    WeaponType.SWORD,
                    WeaponType.MACE,
                    WeaponType.DAGGER,
                    WeaponType.POLEARM,
                    WeaponType.IMPROVISED,
                    WeaponType.THROWN,
                    WeaponType.AXE,
                    WeaponType.BOW
            }, hpScale: 6, manaScale: 3)
        {
            PlayerClassCritChance = 7;
            PurchasePrice = 1500;
            BaseHealth = 20;
            BaseMana = 14;
            HpRegenRate = 2;
            AttackPowerBonus = 2;
            SpellPowerBonus = 1;
        }

        [classmethod]
        public static object FromDB(object cls, object playerClassDict = dict, object player)
        {
            var xp = playerClassDict["xp"];
            var level = Convert.ToInt32(playerClassDict["level"]);
            var playerClass = cls(player, xp, level);
            playerClass.recalculate();
            return playerClass;
        }

        [staticmethod]
        public static object getClassUnlockRequirements()
        {
            return new List<object> {
                    PlayerClassRequirement(Executioner.playerClassId, 10, "Executioner")
                };
        }

        public virtual object prepareAvailableAbilities()
        {
            var unlockedAbilities = new Dictionary<object, object>
            {
            };
            var shout = abilities.Shout();
            unlockedAbilities[shout.magicWord] = shout;
            if (Level >= 5)
            {
                var frenzy = abilities.Frenzy();
                unlockedAbilities[frenzy.magicWord] = frenzy;
            }
            return unlockedAbilities;
        }
    }

    public class Paladin
        : PlayerClass
    {

        public object playerClassId = 7;

        public Paladin(object player, object xp = 0, object level = 1)
            : base(xp, level, player, new List<object> {
                    WeaponType.SWORD,
                    WeaponType.MACE,
                    WeaponType.DAGGER,
                    WeaponType.POLEARM,
                    WeaponType.AXE
            }, hpScale: 5.5, manaScale: 3.5)
        {
            PlayerClassCritChance = 6;
            PurchasePrice = 1500;
            BaseHealth = 16;
            BaseMana = 16;
            HpRegenRate = 2;
            AttackPowerBonus = 1;
            SpellPowerBonus = 1;
        }

        [classmethod]
        public static object FromDB(object cls, object playerClassDict = dict, object player)
        {
            var xp = playerClassDict["xp"];
            var level = Convert.ToInt32(playerClassDict["level"]);
            var playerClass = cls(player, xp, level);
            playerClass.recalculate();
            return playerClass;
        }

        [staticmethod]
        public static object getClassUnlockRequirements()
        {
            return new List<object> {
                    PlayerClassRequirement(Cleric.playerClassId, 10, "Cleric")
                };
        }

        public virtual object prepareAvailableAbilities()
        {
            var unlockedAbilities = new Dictionary<object, object>
            {
            };
            var sacrifice = abilities.Sacrifice();
            unlockedAbilities[sacrifice.magicWord] = sacrifice;
            var heal = abilities.Heal();
            unlockedAbilities[heal.magicWord] = heal;
            if (Level >= 5)
            {
                var fullRestore = abilities.GreaterHeal();
                unlockedAbilities[fullRestore.magicWord] = fullRestore;
            }
            return unlockedAbilities;
        }
    }

    public class MonsterHunter
        : PlayerClass
    {

        public object playerClassId = 8;

        public MonsterHunter(object player, object xp = 0, object level = 1)
            : base(xp, level, player, new List<object> {
                    WeaponType.SWORD,
                    WeaponType.MACE,
                    WeaponType.DAGGER,
                    WeaponType.POLEARM,
                    WeaponType.IMPROVISED,
                    WeaponType.THROWN,
                    WeaponType.AXE,
                    WeaponType.BOW
            }, hpScale: 5, manaScale: 4)
        {
            PlayerClassCritChance = 7;
            PurchasePrice = 1500;
            BaseHealth = 18;
            BaseMana = 16;
            ManaRegenRate = 2;
            AttackPowerBonus = 2;
            SpellPowerBonus = 1;
        }

        [classmethod]
        public static object FromDB(object cls, object playerClassDict = dict, object player)
        {
            var xp = playerClassDict["xp"];
            var level = Convert.ToInt32(playerClassDict["level"]);
            var playerClass = cls(player, xp, level);
            playerClass.recalculate();
            return playerClass;
        }

        [staticmethod]
        public static object getClassUnlockRequirements()
        {
            return new List<object> {
                    PlayerClassRequirement(Ranger.playerClassId, 10, "Ranger")
                };
        }

        public virtual object prepareAvailableAbilities()
        {
            var unlockedAbilities = new Dictionary<object, object>
            {
            };
            var overSizedBearTrap = abilities.OverSizedBearTrap();
            unlockedAbilities[overSizedBearTrap.magicWord] = overSizedBearTrap;
            var turnWeaponToSilver = abilities.TurnWeaponToSilver();
            unlockedAbilities[turnWeaponToSilver.magicWord] = turnWeaponToSilver;
            if (Level >= 5)
            {
                var bigGameTrophy = abilities.BigGameTrophy();
                unlockedAbilities[bigGameTrophy.magicWord] = bigGameTrophy;
            }
            return unlockedAbilities;
        }
    }

    public static object ALLCLASSES = new List<object> {
            Cleric(null, level: 99),
            Ranger(null, level: 99),
            Executioner(null, level: 99),
            Hexer(null, level: 99),
            Mage(null, level: 99),
            Barbarian(null, level: 99),
            Paladin(null, level: 99),
            MonsterHunter(null, level: 99)
        };

    public static object displayInfo(object className = str)
    {
        object text;
        try
        {
            var classToPrint = next(from pc in ALLCLASSES
                                      where pc.name.lower() == className.lower()
                                      select pc);
            text = classToPrint.infoStr();
        }
        catch (StopIteration)
        {
            text = "Could not find class '{className}'";
        }
        return text;
    }
}

File: BossFightBackEnd/boss_fight_py_edition/playerClass.py
import abc
import math
import source.games.boss_fight.ability as abilities
from source.games.boss_fight.FightEnum import WeaponType


class PlayerClassRequirement:
    def __init__(self, class_id: int, lvl_req: int, class_name: str):
        self.class_id = class_id
        self.lvl_req = lvl_req
        self.class_name = class_name

    def __str__(self):
        return f"{self.class_name}:{self.class_id} - lvl {self.lvl_req}"


class PlayerClass(abc.ABC):
    _default_hp_scale = 2
    _default_mana_scale = 2
    player_class_id = -1
    base_health = 10
    base_mana = 10

    def __init__(self, name: str, xp: int, level: int, player, proficient_weapon_types_list: list[WeaponType], hp_scale: float = None, mana_scale: float = None):
        self.name: str = name
        self.xp: int = xp
        self.level: int = level
        self.player = player
        self.abilities: dict = self.prepare_available_abilities()
        self.max_hp: int = 0
        self.max_mana: int = 0
        if hp_scale is not None:
            self.hp_scale: float = hp_scale
        else:
            self.hp_scale: float = self._default_hp_scale
        if mana_scale is not None:
            self.mana_scale: float = mana_scale
        else:
            self.mana_scale: float = self._default_mana_scale
        self.hp_regen_rate: int = 1
        self.mana_regen_rate: int = 1
        self.player_class_crit_chance: int = 0
        self.attack_power_bonus: int = 0
        self.spell_power_bonus: int = 0
        self.class_id: int = self.player_class_id
        self.purchase_price: int = 0
        self.proficient_weapon_types_list: list[WeaponType] = proficient_weapon_types_list
        self.recalculate()

    @abc.abstractmethod
    def prepare_available_abilities(self) -> dict:
        pass

    @classmethod
    @abc.abstractmethod
    def from_dict(cls, player_class_dict: dict, player):
        pass

    @staticmethod
    def get_class_unlock_requirements() -> [PlayerClassRequirement]:
        return []

    def to_dict(self) -> dict:
        player_class_dict = {}
        player_class_dict["name"] = self.name
        player_class_dict["xp"] = self.xp
        player_class_dict["level"] = self.level
        return player_class_dict

    def recalculate(self):
        self.level = self.level
        self.max_hp = math.floor(self.hp_scale * self.level) + self.base_health
        self.max_mana = math.floor(self.mana_scale * self.level) + self.base_mana
        self.abilities: dict = self.prepare_available_abilities()

    def __str__(self):
        return self.name

    def shop_str(self, length_of_longest_player_class_name: int, length_of_longest_player_class_cost: int) -> str:
        purchase_price_str = f"{self.purchase_price:,}"
        purchase_price_str = purchase_price_str.replace(",", ".")
        player_class_req = self.get_class_unlock_requirements()
        player_class_req_str = ", ".join([f"Level {req.lvl_req} {req.class_name}" for req in player_class_req])
        return f"{self.name.ljust(length_of_longest_player_class_name, '.')} {purchase_price_str.ljust(length_of_longest_player_class_cost)} gold {player_class_req_str}"

    def info_str(self) -> str:
        class_name_str = f"Class: {self.name}"
        crit_chance_str = f"Base critical chance: {self.player_class_crit_chance}%"
        base_hp_mana_str = f"Start hp: {self.base_health}; start mana: {self.base_mana}"
        scales_str = f"{self.hp_scale} hp per level; {self.mana_scale} mana per level"
        regen_str = f"{self.hp_regen_rate} hp per regen tick; {self.mana_regen_rate} mana per regen tick"
        proficient_weapons_str = f"Is proficient with: {', '.join([str(p) for p in self.proficient_weapon_types_list])}"
        prev_lvl = self.level
        self.level = 99
        all_spells = self.prepare_available_abilities()
        self.level = prev_lvl
        abilities_str = "Spell list:\n" + f"\n".join([str(s) for s in all_spells.values()])
        return "\n".join([class_name_str, crit_chance_str, base_hp_mana_str, scales_str, regen_str,  abilities_str, proficient_weapons_str])

    def get_health_regen_rate(self) -> int:
        return self.hp_regen_rate

    def get_mana_regen_rate(self) -> int:
        return self.mana_regen_rate

    def get_attack_power_bonus(self) -> int:
        return self.attack_power_bonus

    def get_spell_power_bonus(self) -> int:
        return self.spell_power_bonus

    def level_up(self):
        self.level += 1
        self.xp = 0
        self.recalculate()
        self.player.restore_all_health_and_mana()

    def have_met_requirements(self, player_class_list: ['PlayerClass']) -> bool:
        requirements_met = True
        class_requirements = self.get_class_unlock_requirements()

        for class_req in class_requirements:
            try:
                next(pc for pc in player_class_list if pc.player_class_id == class_req.class_id)
            except StopIteration:
                raise AssertionError(f"You need to unlock class '{class_req.class_name}' first")

            for pc in player_class_list:
                if pc.player_class_id == class_req.class_id:
                    assert pc.level >= class_req.lvl_req, f"Did not meet level requirements for {class_req.class_name}"

        return requirements_met


class Cleric(PlayerClass):
    player_class_id = 1

    def __init__(self, player, xp: int = 0, level: int = 1):
        super(Cleric, self).__init__("Cleric", xp, level, player, [WeaponType.SWORD, WeaponType.MACE, WeaponType.DAGGER, WeaponType.STAFF], hp_scale=1.5, mana_scale=2.5)

    @classmethod
    def from_dict(cls, player_class_dict: dict, player):
        xp = player_class_dict["xp"]
        level = int(player_class_dict["level"])
        player_class = cls(player, xp, level)
        player_class.recalculate()
        return player_class

    def prepare_available_abilities(self) -> dict:
        unlocked_abilities = {}
        heal = abilities.Heal()
        unlocked_abilities[heal.magic_word] = heal
        if self.level >= 5:
            smite = abilities.Smite()
            unlocked_abilities[smite.magic_word] = smite
        return unlocked_abilities


class Executioner(PlayerClass):
    player_class_id = 2

    def __init__(self, player, xp: int = 0, level: int = 1):
        super(Executioner, self).__init__("Executioner", xp, level, player, [WeaponType.SWORD, WeaponType.MACE, WeaponType.DAGGER, WeaponType.POLEARM], hp_scale=2.5, mana_scale=1.5)
        self.player_class_crit_chance = 2

    @classmethod
    def from_dict(cls, player_class_dict: dict, player):
        xp = player_class_dict["xp"]
        level = int(player_class_dict["level"])
        player_class = cls(player, xp, level)
        player_class.recalculate()
        return player_class

    def prepare_available_abilities(self) -> dict:
        unlocked_abilities = {}
        sack_on_head = abilities.SackOnHead()
        unlocked_abilities[sack_on_head.magic_word] = sack_on_head
        if self.level >= 5:
            execute = abilities.Execute()
            unlocked_abilities[execute.magic_word] = execute
        return unlocked_abilities


class Ranger(PlayerClass):
    player_class_id = 3

    def __init__(self, player, xp: int = 0, level: int = 1):
        super(Ranger, self).__init__("Ranger", xp, level, player, [WeaponType.DAGGER, WeaponType.THROWN, WeaponType.BOW, WeaponType.IMPROVISED])
        self.player_class_crit_chance = 4

    @classmethod
    def from_dict(cls, player_class_dict: dict, player):
        xp = player_class_dict["xp"]
        level = int(player_class_dict["level"])
        player_class = cls(player, xp, level)
        player_class.recalculate()
        return player_class

    def prepare_available_abilities(self) -> dict:
        unlocked_abilities = {}
        double_strike = abilities.DoubleStrike()
        unlocked_abilities[double_strike.magic_word] = double_strike
        if self.level >= 5:
            poisoned_bait = abilities.PoisonedBait()
            unlocked_abilities[poisoned_bait.magic_word] = poisoned_bait
        return unlocked_abilities


class Hexer(PlayerClass):
    player_class_id = 4

    def __init__(self, player, xp: int = 0, level: int = 1):
        super(Hexer, self).__init__("Hexer", xp, level, player, [WeaponType.DAGGER, WeaponType.STAFF], hp_scale=3.5, mana_scale=4)
        self.player_class_crit_chance = 3
        self.purchase_price = 1000
        self.base_health = 12
        self.base_mana = 13
        self.mana_regen_rate = 2
        self.attack_power_bonus = 1

    @classmethod
    def from_dict(cls, player_class_dict: dict, player):
        xp = player_class_dict["xp"]
        level = int(player_class_dict["level"])
        player_class = cls(player, xp, level)
        player_class.recalculate()
        return player_class

    def prepare_available_abilities(self) -> dict:
        unlocked_abilities = {}
        hex = abilities.Hex()
        unlocked_abilities[hex.magic_word] = hex
        if self.level >= 5:
            fracture_skin = abilities.FractureSkin()
            unlocked_abilities[fracture_skin.magic_word] = fracture_skin
        return unlocked_abilities


class Mage(PlayerClass):
    player_class_id = 5

    def __init__(self, player, xp: int = 0, level: int = 1):
        super(Mage, self).__init__("Mage", xp, level, player, [WeaponType.DAGGER, WeaponType.STAFF], hp_scale=3, mana_scale=4.5)
        self.player_class_crit_chance = 4
        self.purchase_price = 1000
        self.base_health = 11
        self.base_mana = 14
        self.mana_regen_rate = 2
        self.spell_power_bonus = 1

    @classmethod
    def from_dict(cls, player_class_dict: dict, player):
        xp = player_class_dict["xp"]
        level = int(player_class_dict["level"])
        player_class = cls(player, xp, level)
        player_class.recalculate()
        return player_class

    def prepare_available_abilities(self) -> dict:
        unlocked_abilities = {}
        ignite = abilities.Ignite()
        unlocked_abilities[ignite.magic_word] = ignite
        if self.level >= 5:
            enchant_weapon = abilities.EnchantWeapon()
            unlocked_abilities[enchant_weapon.magic_word] = enchant_weapon
        return unlocked_abilities


class Barbarian(PlayerClass):
    player_class_id = 6

    def __init__(self, player, xp: int = 0, level: int = 1):
        super(Barbarian, self).__init__("Barbarian", xp, level, player, [WeaponType.SWORD, WeaponType.MACE, WeaponType.DAGGER, WeaponType.POLEARM, WeaponType.IMPROVISED, WeaponType.THROWN, WeaponType.AXE, WeaponType.BOW], hp_scale=6, mana_scale=3)
        self.player_class_crit_chance = 7
        self.purchase_price = 1500
        self.base_health = 20
        self.base_mana = 14
        self.hp_regen_rate = 2
        self.attack_power_bonus = 2
        self.spell_power_bonus = 1

    @classmethod
    def from_dict(cls, player_class_dict: dict, player):
        xp = player_class_dict["xp"]
        level = int(player_class_dict["level"])
        player_class = cls(player, xp, level)
        player_class.recalculate()
        return player_class

    @staticmethod
    def get_class_unlock_requirements() -> [PlayerClassRequirement]:
        return [PlayerClassRequirement(Executioner.player_class_id, 10, "Executioner")]

    def prepare_available_abilities(self) -> dict:
        unlocked_abilities = {}
        shout = abilities.Shout()
        unlocked_abilities[shout.magic_word] = shout
        if self.level >= 5:
            frenzy = abilities.Frenzy()
            unlocked_abilities[frenzy.magic_word] = frenzy
        return unlocked_abilities


class Paladin(PlayerClass):
    player_class_id = 7

    def __init__(self, player, xp: int = 0, level: int = 1):
        super(Paladin, self).__init__("Paladin", xp, level, player, [WeaponType.SWORD, WeaponType.MACE, WeaponType.DAGGER, WeaponType.POLEARM, WeaponType.AXE], hp_scale=5.5, mana_scale=3.5)
        self.player_class_crit_chance = 6
        self.purchase_price = 1500
        self.base_health = 16
        self.base_mana = 16
        self.hp_regen_rate = 2
        self.attack_power_bonus = 1
        self.spell_power_bonus = 1

    @classmethod
    def from_dict(cls, player_class_dict: dict, player):
        xp = player_class_dict["xp"]
        level = int(player_class_dict["level"])
        player_class = cls(player, xp, level)
        player_class.recalculate()
        return player_class

    @staticmethod
    def get_class_unlock_requirements() -> [PlayerClassRequirement]:
        return [PlayerClassRequirement(Cleric.player_class_id, 10, "Cleric")]

    def prepare_available_abilities(self) -> dict:
        unlocked_abilities = {}
        sacrifice = abilities.Sacrifice()
        unlocked_abilities[sacrifice.magic_word] = sacrifice
        heal = abilities.Heal()
        unlocked_abilities[heal.magic_word] = heal
        if self.level >= 5:
            full_restore = abilities.GreaterHeal()
            unlocked_abilities[full_restore.magic_word] = full_restore
        return unlocked_abilities


class MonsterHunter(PlayerClass):
    player_class_id = 8

    def __init__(self, player, xp: int = 0, level: int = 1):
        super(MonsterHunter, self).__init__("Monster Hunter", xp, level, player, [WeaponType.SWORD, WeaponType.MACE, WeaponType.DAGGER, WeaponType.POLEARM, WeaponType.IMPROVISED, WeaponType.THROWN, WeaponType.AXE, WeaponType.BOW], hp_scale=5, mana_scale=4)
        self.player_class_crit_chance = 7
        self.purchase_price = 1500
        self.base_health = 18
        self.base_mana = 16
        self.mana_regen_rate = 2
        self.attack_power_bonus = 2
        self.spell_power_bonus = 1

    @classmethod
    def from_dict(cls, player_class_dict: dict, player):
        xp = player_class_dict["xp"]
        level = int(player_class_dict["level"])
        player_class = cls(player, xp, level)
        player_class.recalculate()
        return player_class

    @staticmethod
    def get_class_unlock_requirements() -> [PlayerClassRequirement]:
        return [PlayerClassRequirement(Ranger.player_class_id, 10, "Ranger")]

    def prepare_available_abilities(self) -> dict:
        unlocked_abilities = {}
        over_sized_bear_trap = abilities.OverSizedBearTrap()  # Monster misses next attack, starts bleeding, and takes upfront damage
        unlocked_abilities[over_sized_bear_trap.magic_word] = over_sized_bear_trap
        turn_weapon_to_silver = abilities.TurnWeaponToSilver()  # deals extra damage to exotic monster types
        unlocked_abilities[turn_weapon_to_silver.magic_word] = turn_weapon_to_silver
        if self.level >= 5:
            big_game_trophy = abilities.BigGameTrophy()  # Extra dmg, gain bounty on kill, easier to crit monster
            unlocked_abilities[big_game_trophy.magic_word] = big_game_trophy
        return unlocked_abilities


ALL_CLASSES: list[PlayerClass] = [Cleric(None, level=99), Ranger(None, level=99), Executioner(None, level=99), Hexer(None, level=99), Mage(None, level=99), Barbarian(None, level=99), Paladin(None, level=99), MonsterHunter(None, level=99)]


def display_info(class_name: str) -> str:
    try:
        class_to_print = next(pc for pc in ALL_CLASSES if pc.name.lower() == class_name.lower())
        text = class_to_print.info_str()
    except StopIteration:
        text = f"Could not find class '{class_name}'"
    return text
